velox/functions/sparksql/aggregates/FirstLastAggregate.cpp

```diff
@@ -0,0 +1,342 @@
+      }
+
+      if constexpr (!ignoreNull) {
+        exec::Aggregate::setNull(group);
+        *accumulator = TDataType();
+      }
+    }
+  }
+
+  void updateNonNumeric(vector_size_t i, char* group, DecodedVector& decoded) {
+    auto accumulator = exec::Aggregate::value<TAccumulator>(group);
+
+    if (!decoded.isNullAt(i)) {
+      exec::Aggregate::clearNull(group);
+      *accumulator = SingleValueAccumulator();
+      accumulator->value().write(
+          decoded.base(), decoded.index(i), exec::Aggregate::allocator_);
+      return;
+    }
+
+    if constexpr (!ignoreNull) {
+      exec::Aggregate::setNull(group);
+      *accumulator = SingleValueAccumulator();
+    }
+  }
+};
+
+} // namespace
+
+template <template <bool B1, typename T, bool B2> class TClass, bool ignoreNull>
+bool registerFirstLast(const std::string& name) {
+  std::vector<std::shared_ptr<exec::AggregateFunctionSignature>> signatures = {
+      exec::AggregateFunctionSignatureBuilder()
+          .typeVariable("T")
+          .argumentType("T")
+          .intermediateType("T")
+          .returnType("T")
+          .build()};
+
+  return exec::registerAggregateFunction(
+      name,
+      std::move(signatures),
+      [name](
+          core::AggregationNode::Step step,
+          const std::vector<TypePtr>& argTypes,
+          const TypePtr& resultType) -> std::unique_ptr<exec::Aggregate> {
+        VELOX_CHECK_EQ(argTypes.size(), 1, "{} takes only 1 arguments", name);
+        const auto& inputType = argTypes[0];
+        TypeKind dataKind = inputType->kind();
+        switch (dataKind) {
+          case TypeKind::BOOLEAN:
+            return std::make_unique<TClass<ignoreNull, bool, true>>(resultType);
+          case TypeKind::TINYINT:
+            return std::make_unique<TClass<ignoreNull, int8_t, true>>(
+                resultType);
+          case TypeKind::SMALLINT:
+            return std::make_unique<TClass<ignoreNull, int16_t, true>>(
+                resultType);
+          case TypeKind::INTEGER:
+            return std::make_unique<TClass<ignoreNull, int32_t, true>>(
+                resultType);
+          case TypeKind::BIGINT:
+            return std::make_unique<TClass<ignoreNull, int64_t, true>>(
+                resultType);
+          case TypeKind::REAL:
+            return std::make_unique<TClass<ignoreNull, float, true>>(
+                resultType);
+          case TypeKind::DOUBLE:
+            return std::make_unique<TClass<ignoreNull, double, true>>(
+                resultType);
+          case TypeKind::TIMESTAMP:
+            return std::make_unique<TClass<ignoreNull, Timestamp, true>>(
+                resultType);
+          case TypeKind::DATE:
+            return std::make_unique<TClass<ignoreNull, Date, true>>(resultType);
+          case TypeKind::VARCHAR:
+          case TypeKind::ARRAY:
+          case TypeKind::MAP:
+          case TypeKind::ROW:
+            // int8_t here is a tricky for compile, has no meaning.
```

Comment:
This comment is confusing to me. What does it mean? Consider using ComplexType instead of int8_t (see Type.h).